src/libstore/build-capability.hh

```diff
@@ -0,0 +1,78 @@
+#pragma once
+///@file
+
+#include <optional>
+#include <set>
+#include <string>
+
+namespace nix {
+
+class Schedulable {
+public:
+    virtual std::string_view getSystem() const = 0;
+    virtual const std::set<std::string> & getRequiredFeatures() const = 0;
+    virtual bool getPreferLocalBuild() const = 0;
+};
+
+/**
+ * Parameters that determine which derivations can be built.
+ *
+ * *Where* it can be built is determined by context.
+ */
+struct BuildCapability {
+    /**
+     * For a derivation to be buildable by this capability, `system` must match the derivation `system` by case sensitive string equality.
+     *
+     * In a given context, multiple `system`s may be supported. This is represented by having multiple `BuildCapability`s.
+     */
+    std::string system;
+
+    /**
+     * For a derivation to be buildable by this capability, `supportedFeatures` must be a superset of the derivation's `requiredFeatures`, or be equal.
+     */
+    std::set<std::string> supportedFeatures;
+
+    /**
+     * For a derivation to be buildable by this capability, `mandatoryFeatures` must be a subset of the derivation's `requiredFeatures`, or be equal.
+     */
+    std::set<std::string> mandatoryFeatures;
+
+    bool canBuild(const Schedulable & schedulable) const;
+};
+
+/**
+ * Extends `BuildCapability` to include scheduling information.
+ */
+struct SchedulableCapability {
+    /**
+     * Which derivations can be built.
+     */
+    BuildCapability capability;
```

Comment:
Should be a list of capabilities, because of shared resources? Alternatively, see https://github.com/NixOS/nix/pull/9899/files#r1478444104

Do we still conflate too many concepts? Maybe we need a build routing tree that's about available resources (and load). Each node may or may not impose restrictions on builds that are routed through it. Each node may or may not have information about how to contact a builder.